zebra/zebra_evpn_mac.c

```diff
@@ -1323,7 +1323,7 @@ int zebra_evpn_mac_send_del_to_client(vni_t vni, struct ethaddr *macaddr,
  */
 struct hash *zebra_mac_db_create(const char *desc)
 {
-	return hash_create(mac_hash_keymake, mac_cmp, desc);
+	return hash_create_size(8, mac_hash_keymake, mac_cmp, desc);
```

Comment:
I'm not crazy about the sort-of-magic value "8" - could we have "FRR_HASH_SMALL_SIZE" or something like that?

Reply:
```sharpd@eva ~/frr5 (hash_size_reduction)> git grep hash_create_size
bgpd/bgp_aspath.c:      ashash = hash_create_size(32768, aspath_key_make, aspath_cmp,
bgpd/bgp_clist.c:               hash_create_size(4, bgp_clist_hash_key_community_list,
bgpd/bgp_clist.c:               hash_create_size(4, bgp_clist_hash_key_community_list,
bgpd/bgp_clist.c:               hash_create_size(4, bgp_clist_hash_key_community_list,
bgpd/bgp_keepalives.c:#include "hash.h"         // for hash, hash_clean, hash_create_size...
bgpd/bgp_keepalives.c:  peerhash = hash_create_size(2048, peer_hash_key, peer_hash_cmp, NULL);
bgpd/bgp_pbr.c:         hash_create_size(8, bgp_pbr_match_hash_key,
bgpd/bgp_pbr.c:         hash_create_size(8, bgp_pbr_action_hash_key,
bgpd/bgp_pbr.c:         hash_create_size(8, bgp_pbr_rule_hash_key,
bgpd/bgp_pbr.c:         bpm->entry_hash = hash_create_size(8,
lib/command.c:  node->cmd_hash = hash_create_size(16, cmd_hash_key, cmd_hash_cmp,
lib/hash.c:struct hash *hash_create_size(unsigned int size,
lib/hash.c:     return hash_create_size(HASH_INITIAL_SIZE, hash_key, hash_cmp, name);
lib/hash.h:hash_create_size(unsigned int size, unsigned int (*hash_key)(const void *),
lib/if_rmap.c:  ctx->ifrmaphash = hash_create_size(4, if_rmap_hash_make, if_rmap_hash_cmp,
lib/routemap.c:         hash_create_size(8, route_map_dep_data_hash_make_key,
lib/routemap.c:         hash_create_size(8, route_map_hash_key_make, route_map_hash_cmp,
lib/routemap.c:         route_map_dep_hash[i] = hash_create_size(
lib/thread.c:   rv->cpu_record = hash_create_size(
lib/vrf.c:      return hash_create_size(32, vrf_hash_bitmap_key, vrf_hash_bitmap_cmp,
pbrd/pbr_nht.c: new->nhh = hash_create_size(8, pbr_nh_hash_key, pbr_nh_hash_equal,
pbrd/pbr_nht.c: pbr_nhg_hash = hash_create_size(
pbrd/pbr_nht.c:         hash_create_size(16, (unsigned int (*)(const void *))nexthop_hash,
pimd/pim_instance.c:    pim->rpf_hash = hash_create_size(256, pim_rpf_hash_key, pim_rpf_equal,
pimd/pim_nht.c: pnc->upstream_hash = hash_create_size(8192, pim_upstream_hash_key,
zebra/zebra_evpn_mac.c: return hash_create_size(8, mac_hash_keymake, mac_cmp, desc);
zebra/zebra_evpn_neigh.c:       return hash_create_size(8, neigh_hash_keymake, neigh_cmp, desc);
zebra/zebra_mpls.c:             hash_create_size(8, label_hash, label_cmp, buffer);
zebra/zebra_mpls.c:     zvrf->lsp_table = hash_create_size(8, label_hash, label_cmp, buffer);
zebra/zebra_router.c:   zrouter.rules_hash = hash_create_size(8, zebra_pbr_rules_hash_key,
zebra/zebra_router.c:           hash_create_size(8, zebra_pbr_ipset_hash_key,
zebra/zebra_router.c:   zrouter.ipset_entry_hash = hash_create_size(
zebra/zebra_router.c:   zrouter.iptable_hash = hash_create_size(8, zebra_pbr_iptable_hash_key,
zebra/zebra_router.c:           hash_create_size(8, zebra_nhg_hash_key, zebra_nhg_hash_equal,
zebra/zebra_router.c:           hash_create_size(8, zebra_nhg_id_key, zebra_nhg_hash_id_equal,
zebra/zebra_vxlan.c:    zvrf->evpn_table = hash_create_size(8, zebra_evpn_hash_keymake,
zebra/zebra_vxlan.c:    zvrf->vxlan_sg_table = hash_create_size(8, zebra_vxlan_sg_hash_key_make,
```

Reply:
> I'm not crazy about the sort-of-magic value "8"

In this case it's not really a magic value. A magic value to me is a value where it's not immediately clear what the significance of the number is. In this case, it's always clear, you're saying "I think this hash needs about N items". A define probably hurts clarity in this case.

Reply:
Sure, just something that occurred to me when I looked at it. I agree that it's clearer to see that actual value than to have to look up some constant in an .h file...

Reply:
I'm not going to set the initial bucket size to 16.  hash expansion does not even begin to matter until we see expansions for hash sizes of 1k.  On the other hand 16*32*3k is very different in memory requirements for 8*32*3k.

It might actually be worth our while to create an event to expand the hash instead of doing it inline with hash_get.

Reply:
> It might actually be worth our while to create an event to expand the hash instead of doing it inline with hash_get.

I've been thinking about that too & have it on my "long term maybe TODO".  But it's not quite clear cut;  if it's just something we do in the same thread but later, there's not much of a real benefit, and if it's in a separate thread it needs some kind of synchronization…